File: src/lib/wow/export/export-progress.ts
/**
 * In-process export progress tracking for ADT batch exports.
 * Shared between wow-data-server (writer) and the web UI server (reader via REST).
 */
import type { ADTExportOptions } from './adt/map-export-utils';

/** Legacy OBJ export still advances per model; keep a generous cap. */
export const MODEL_STEP_BUDGET = 512;

export const FIXED_OVERHEAD_STEPS = 2;

export interface ExportProgressSnapshot {
  completedSteps: number;
  totalSteps: number;
  tileIndex: number;
  tileCount: number;
  stepsPerTile: number;
  currentTile?: { x: number; y: number };
  taskName?: string;
  taskValue?: number;
  taskMax?: number;
}

export interface ExportProgress {
  advance(steps?: number): void;
  setLabel(name: string, value?: number, max?: number): void;
  syncTileComplete(): void;
}

/**
 * Fixed phase budget per tile for a uniform tiles × stepsPerTile progress bar.
 * Direct mode uses coarse phases (typically 6 steps/tile); legacy OBJ export
 * keeps the per-model step budget.
 */
export function computeStepsPerTile(quality: number, options: ADTExportOptions): number {
  let steps = FIXED_OVERHEAD_STEPS;
  if (quality !== 0) steps += 1;
  if (options.mapsIncludeM2 || options.mapsIncludeWMO || options.mapsIncludeGameObjects) {
    steps += options.mapsDirectModels ? 1 : MODEL_STEP_BUDGET;
  }
  if (options.mapsIncludeLiquid) steps += 1;
  if (options.mapsIncludeFoliage) steps += 1;
  return steps;
}

const progressStore = new Map<string, ExportProgressSnapshot>();

export function getExportProgressSnapshot(key: string): ExportProgressSnapshot | undefined {
  return progressStore.get(key);
}

export function clearExportProgressSnapshot(key: string): void {
  progressStore.delete(key);
}

export interface BatchExportProgressParams {
  key: string;
  tileIndex: number;
  tileCount: number;
  stepsPerTile: number;
  currentTile: { x: number; y: number };
}

export function createBatchExportProgress(params: BatchExportProgressParams): ExportProgress {
  const {
    key, tileIndex, tileCount, stepsPerTile, currentTile,
  } = params;
  const totalSteps = tileCount * stepsPerTile;
  const tileBase = tileIndex * stepsPerTile;

  let tileSteps = 0;
  let taskName = '';
  let taskValue: number | undefined;
  let taskMax: number | undefined;

  const publish = (completedSteps: number): void => {
    progressStore.set(key, {
      completedSteps,
      totalSteps,
      tileIndex,
      tileCount,
      stepsPerTile,
      currentTile,
      taskName: taskName || undefined,
      taskValue,
      taskMax,
    });
  };

  publish(tileBase);

  return {
    advance(steps = 1) {
      tileSteps = Math.min(tileSteps + steps, Math.max(0, stepsPerTile - 1));
      publish(tileBase + tileSteps);
    },
    setLabel(name: string, value?: number, max?: number) {
      taskName = name;
      taskValue = value;
      taskMax = max;
      publish(tileBase + tileSteps);
    },
    syncTileComplete() {
      tileSteps = stepsPerTile;
      publish(tileBase + stepsPerTile);
    },
  };
}

export function finalizeExportProgress(key: string): void {
  const snap = progressStore.get(key);
  if (snap) {
    progressStore.set(key, {
      ...snap,
      completedSteps: snap.totalSteps,
      taskName: 'Complete',
      taskValue: undefined,
      taskMax: undefined,
    });
  }
}

File: golang/internal/wow/export/export_progress.go
package export

import (
	"sync"

	apicasc "github.com/pqhuy98/wow-converter/internal/wow/casc"
)

const FixedOverheadSteps = 2

// ADTExportOptions are normalized ADT export settings.
type ADTExportOptions struct {
	PathFormat             string
	EnableSharedTextures   bool
	OverwriteFiles         bool
	SplitAlphaMaps         bool
	SplitLargeTerrainBakes bool
	MapsIncludeHoles       bool
	EnableSharedChildren   bool
	EnableAbsoluteCSVPaths bool
	ModelsExportCollision  bool
	MapsIncludeWMO         bool
	MapsIncludeM2          bool
	MapsIncludeWMOSets     bool
	ExportFoliageMeta      bool
	MapsIncludeFoliage     bool
	MapsIncludeLiquid      bool
	MapsIncludeGameObjects bool
}

// ComputeStepsPerTile returns the fixed phase budget per tile.
func ComputeStepsPerTile(quality int, options ADTExportOptions) int {
	steps := FixedOverheadSteps
	if quality != 0 {
		steps++
	}
	if options.MapsIncludeM2 || options.MapsIncludeWMO || options.MapsIncludeGameObjects {
		steps++
	}
	if options.MapsIncludeLiquid {
		steps++
	}
	if options.MapsIncludeFoliage {
		steps++
	}
	return steps
}

var progressStore sync.Map

// BatchExportProgressParams configures batch export progress tracking.
type BatchExportProgressParams struct {
	Key          string
	TileIndex    int
	TileCount    int
	StepsPerTile int
	CurrentTile  apicasc.TileCoord
}

// ProgressReporter tracks export progress for a single tile.
type ProgressReporter struct {
	key, taskName              string
	tileIndex, tileCount       int
	stepsPerTile, tileSteps    int
	totalSteps, tileBase       int
	currentTile                apicasc.TileCoord
	taskValue, taskMax         *int
}

// CreateBatchExportProgress creates a progress reporter.
func CreateBatchExportProgress(params BatchExportProgressParams) *ProgressReporter {
	totalSteps := params.TileCount * params.StepsPerTile
	tileBase := params.TileIndex * params.StepsPerTile
	p := &ProgressReporter{
		key: params.Key, tileIndex: params.TileIndex, tileCount: params.TileCount,
		stepsPerTile: params.StepsPerTile, totalSteps: totalSteps, tileBase: tileBase,
		currentTile: params.CurrentTile,
	}
	p.publish(tileBase)
	return p
}

func (p *ProgressReporter) publish(completed int) {
	snap := apicasc.ExportProgressSnapshot{
		CompletedSteps: completed,
		TotalSteps:     p.totalSteps,
		TileIndex:      p.tileIndex,
		TileCount:      p.tileCount,
		StepsPerTile:   p.stepsPerTile,
		CurrentTile:    &p.currentTile,
	}
	if p.taskName != "" {
		snap.TaskName = p.taskName
	}
	if p.taskValue != nil {
		snap.TaskValue = *p.taskValue
	}
	if p.taskMax != nil {
		snap.TaskMax = *p.taskMax
	}
	progressStore.Store(p.key, snap)
}

// Advance increments completed steps within the current tile.
func (p *ProgressReporter) Advance(steps ...int) {
	n := 1
	if len(steps) > 0 {
		n = steps[0]
	}
	max := p.stepsPerTile - 1
	if max < 0 {
		max = 0
	}
	p.tileSteps += n
	if p.tileSteps > max {
		p.tileSteps = max
	}
	p.publish(p.tileBase + p.tileSteps)
}

// SetLabel updates the current task label.
func (p *ProgressReporter) SetLabel(name string, value ...int) {
	p.taskName = name
	if len(value) > 0 {
		v := value[0]
		p.taskValue = &v
	}
	if len(value) > 1 {
		v := value[1]
		p.taskMax = &v
	}
	p.publish(p.tileBase + p.tileSteps)
}

// SyncTileComplete marks the tile as fully complete.
func (p *ProgressReporter) SyncTileComplete() {
	p.tileSteps = p.stepsPerTile
	p.publish(p.tileBase + p.stepsPerTile)
}

// GetExportProgressSnapshot returns progress for a key.
func GetExportProgressSnapshot(key string) (*apicasc.ExportProgressSnapshot, bool) {
	if v, ok := progressStore.Load(key); ok {
		snap := v.(apicasc.ExportProgressSnapshot)
		return &snap, true
	}
	return nil, false
}

// FinalizeExportProgress marks a batch export complete.
func FinalizeExportProgress(key string) {
	if v, ok := progressStore.Load(key); ok {
		snap := v.(apicasc.ExportProgressSnapshot)
		snap.CompletedSteps = snap.TotalSteps
		snap.TaskName = "Complete"
		snap.TaskValue = 0
		snap.TaskMax = 0
		progressStore.Store(key, snap)
	}
}

// ClearExportProgressSnapshot removes a progress entry.
func ClearExportProgressSnapshot(key string) {
	progressStore.Delete(key)
}

// ExportProgressService implements casc.ExportProgress.
type ExportProgressService struct{}

func (ExportProgressService) GetSnapshot(key string) (*apicasc.ExportProgressSnapshot, bool) {
	return GetExportProgressSnapshot(key)
}

func (ExportProgressService) Finalize(key string) (*apicasc.ExportProgressSnapshot, bool) {
	FinalizeExportProgress(key)
	return GetExportProgressSnapshot(key)
}
